### ISSUE TITLE
restsearch - galaxy search filters

### ISSUE THREAD

**Issue by iglocska** [MISP/MISP#3644]
search by element (key : value)
search by cluster name
search by alias name

**Comment by StefanKelm** [comment]
Bump.

It would also be really cool (TM) to be able to use the GUI to search for things like "all events containing tags from the _Malpedia_ cluster"

**Comment by mokaddem** [comment]
I think the issue could be closed as it is now possible to search with the following parameters:

**galaxy.`element_key`: [`element_value(s)`]**
```
/attributes/restsearch/
{
    "galaxy.cfr-suspected-victims": ["China", "Japan"],
    "galaxy.cfr-target-category" : ["Government"],
    "galaxy.synonyms": "APT29",
}
```

And same with organisation,
**orgc.`element_key`: [`element_value(s)`]**
```
/events/restsearch/
{
    "orgc.nationality": ["Hungary", "Belgium"]
}
```

@StefanKelm, you can already do it with RestSearch by passing the following parameter:
```
{
    "tags": "misp-galaxy:malpedia=%"
}
```

**Comment by iglocska** [comment]
Absolutely, that's exactly what we needed. Though @StefanKelm's reply highlights something we're still missing - opening up the searches in the GUI to these filters too.

**Comment by mokaddem** [comment]
If the GUI means the query builder, it is not possible unfortunately.
The issue is `element_key` can be anything. However the rules in te QB requires to be defined on initialisation.

The only possible way is to manually edit the JSON body by hand

**Comment by iglocska** [comment]
I meant the attribute search and the event filters, not the rest client.

**Comment by StefanKelm** [comment]
You read my mind...  ;-)

**Comment by Wachizungu** [comment]
Not sure whether it is in scope of this issue, but being able to search clusters based on cluster element key-value pairs would be very valuable.

Use case example: Find all events associated with a threat actor linked to nation x (or just clusters).

2 steps:
1. Cluster search: Galaxy="threat actor galaxy" AND cluster element nation="x".
2. Events search: based on clusters found in step 1

As far as I can see this type of AND wouldn't be possible otherwise.

**Comment by iglocska** [comment]
Made some changes to open it up to the galaxy cluster restSearch. Whilst we don't allow AND/OR operations, all parameter keys are ANDed by default and all values within the same key are ORed.

Example:

```
{
    "value": ["Sofacy", "%APT-29%"],
    "elements": [
        "country": ["ru", "hu"]
    ]
}
```

Would return all galaxy clusters named Sofacy or a name containing apt-29 which carry the country:ru or country:hu elements.

**Comment by Wachizungu** [comment]
Clarification...
elements should be a dict, not array. Country key values are typically uppercase so an example is:
```
{
    "galaxy_uuid": "698774c7-8022-42c4-917f-8d6e4f06ada3",
    "elements": {
        "country": ["RU", "CN"]
    }
}
```

**Comment by adulau** [comment]
@Wachizungu Something related, should we review the threat-actor galaxy concerning the country codes used?
